zcash_primitives/src/transaction/builder.rs

```diff
@@ -261,6 +263,46 @@ impl TransparentInputs {
     fn apply_signatures(&self, _: &mut TransactionData, _: consensus::BranchId) {}
 }
 
+struct TzeInputInfo<'a, BuildCtx> {
+    prevout: TzeOut,
+    builder: Box<dyn FnOnce(&BuildCtx) -> Result<TzeIn, Error> + 'a>,
+}
+
+struct TzeInputs<'a, BuildCtx> {
+    builders: Vec<TzeInputInfo<'a, BuildCtx>>,
+}
+
+impl<'a, BuildCtx> TzeInputs<'a, BuildCtx> {
+    fn default() -> Self {
+        TzeInputs { builders: vec![] }
+    }
+
+    fn push<WBuilder, W: ToPayload>(
+        &mut self,
+        extension_id: u32,
+        prevout: (OutPoint, TzeOut),
+        builder: WBuilder,
+    ) where
+        WBuilder: 'a + FnOnce(&BuildCtx) -> Result<W, Error>,
+    {
+        let (outpoint, tzeout) = prevout;
+        self.builders.push(TzeInputInfo {
+            prevout: tzeout,
+            builder: Box::new(move |ctx| {
+                let (mode, payload) = builder(&ctx).map(|x| x.to_payload())?;
+                Ok(TzeIn {
```

Comment:
I'm not sure I understand. Builder::add_tze_input defers any construction of the witness to this point, so that the witness could commit internally to the binding signatures produced on line 746. The context value ends up being the entire mtx.